File: golang/pkg/agentbay/beta_skills.go
package agentbay

import (
	"fmt"
	"strings"
	"time"

	"github.com/alibabacloud-go/tea/tea"
	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
)

// SkillMetadataItem represents official skill metadata returned by GetSkillMetaData or ListSkillMetaData.
type SkillMetadataItem struct {
	Name        string
	Description string
}

// SkillsMetadataResult contains skills list and root path from GetSkillMetaData.
type SkillsMetadataResult struct {
	Skills         []SkillMetadataItem
	SkillsRootPath string
}

// GetMetadataOptions provides optional filtering for GetMetadata.
type GetMetadataOptions struct {
	ImageID    string
	SkillNames []string
}

// BetaSkillsService provides beta methods to list and get official skills metadata.
type BetaSkillsService struct {
	AgentBay *AgentBay
}

// GetMetadata fetches skills metadata via POP Action GetSkillMetaData.
// Returns SkillsMetadataResult with skills list and skillsRootPath.
func (s *BetaSkillsService) GetMetadata(opts ...GetMetadataOptions) (*SkillsMetadataResult, error) {
	opt := GetMetadataOptions{}
	if len(opts) > 0 {
		opt = opts[0]
	}

	request := &mcp.GetSkillMetaDataRequest{
		Authorization: tea.String("Bearer " + s.AgentBay.APIKey),
	}
	if opt.ImageID != "" {
		request.ImageId = tea.String(opt.ImageID)
	}
	if len(opt.SkillNames) > 0 {
		skillPtrs := make([]*string, len(opt.SkillNames))
		for i, name := range opt.SkillNames {
			skillPtrs[i] = tea.String(name)
		}
		request.SkillGroupIds = skillPtrs
	}

	maxAttempts := 3
	delay := 200 * time.Millisecond
	var resp *mcp.GetSkillMetaDataResponse
	var err error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		resp, err = s.AgentBay.Client.GetSkillMetaData(request)
		if err == nil {
			break
		}
		if attempt < maxAttempts {
			errStr := err.Error()
			if strings.Contains(errStr, "ServiceUnavailable") || strings.Contains(errStr, "503") {
				time.Sleep(delay)
				delay *= 2
				continue
			}
		}
	}
	if err != nil {
		return nil, fmt.Errorf("GetSkillMetaData failed: %w", err)
	}
	if resp == nil || resp.Body == nil {
		return nil, fmt.Errorf("GetSkillMetaData failed: missing response body")
	}

	return ParseGetSkillMetaDataResponse(resp.Body)
}

// ParseGetSkillMetaDataResponse parses GetSkillMetaDataResponseBody into SkillsMetadataResult.
// Exported for unit testing.
func ParseGetSkillMetaDataResponse(body *mcp.GetSkillMetaDataResponseBody) (*SkillsMetadataResult, error) {
	if body == nil {
		return nil, fmt.Errorf("GetSkillMetaData failed: missing response body")
	}
	if body.Success == nil || !*body.Success {
		msg := ""
		if body.Message != nil {
			msg = *body.Message
		}
		code := ""
		if body.Code != nil {
			code = *body.Code
		}
		if code != "" {
			return nil, fmt.Errorf("GetSkillMetaData failed: [%s] %s", code, msg)
		}
		if msg == "" {
			msg = "Unknown error"
		}
		return nil, fmt.Errorf("GetSkillMetaData failed: %s", msg)
	}

	data := body.Data
	if data == nil {
		return nil, fmt.Errorf("GetSkillMetaData failed: invalid Data field")
	}

	skillPath := ""
	if data.SkillPath != nil {
		skillPath = *data.SkillPath
	}

	items := make([]SkillMetadataItem, 0)
	if data.MetaDataList != nil {
		for _, raw := range data.MetaDataList {
			if raw == nil {
				continue
			}
			name := strings.TrimSpace(tea.StringValue(raw.Name))
			if name == "" {
				continue
			}
			desc := tea.StringValue(raw.Description)
			items = append(items, SkillMetadataItem{
				Name:        name,
				Description: desc,
			})
		}
	}

	return &SkillsMetadataResult{
		Skills:         items,
		SkillsRootPath: skillPath,
	}, nil
}

// ListMetadata lists official skills metadata via POP Action ListSkillMetaData.
//
// Deprecated: Use GetMetadata instead, which returns both skills and skillsRootPath.
func (s *BetaSkillsService) ListMetadata() ([]SkillMetadataItem, error) {
	request := &mcp.ListSkillMetaDataRequest{
		Authorization: tea.String("Bearer " + s.AgentBay.APIKey),
	}

	maxAttempts := 3
	delay := 200 * time.Millisecond
	var resp *mcp.ListSkillMetaDataResponse
	var err error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		resp, err = s.AgentBay.Client.ListSkillMetaData(request)
		if err == nil {
			break
		}
		if attempt < maxAttempts {
			errStr := err.Error()
			if strings.Contains(errStr, "ServiceUnavailable") || strings.Contains(errStr, "503") {
				time.Sleep(delay)
				delay *= 2
				continue
			}
		}
	}
	if err != nil {
		return nil, fmt.Errorf("ListSkillMetaData failed: %w", err)
	}
	if resp == nil || resp.Body == nil {
		return nil, fmt.Errorf("ListSkillMetaData failed: missing response body")
	}

	body := resp.Body
	if body.Success == nil || !*body.Success {
		msg := ""
		if body.Message != nil {
			msg = *body.Message
		}
		code := ""
		if body.Code != nil {
			code = *body.Code
		}
		if code != "" {
			return nil, fmt.Errorf("ListSkillMetaData failed: [%s] %s", code, msg)
		}
		if msg == "" {
			msg = "Unknown error"
		}
		return nil, fmt.Errorf("ListSkillMetaData failed: %s", msg)
	}

	data := body.Data
	if data == nil {
		return nil, fmt.Errorf("ListSkillMetaData failed: invalid Data field")
	}

	items := make([]SkillMetadataItem, 0, len(data))
	for _, raw := range data {
		if raw == nil {
			continue
		}
		name := strings.TrimSpace(tea.StringValue(raw.Name))
		if name == "" {
			continue
		}
		desc := tea.StringValue(raw.Description)
		items = append(items, SkillMetadataItem{
			Name:        name,
			Description: desc,
		})
	}
	return items, nil
}

File: typescript/src/beta-skills.ts
import { AgentBay } from "./agent-bay";
import * as $_client from "./api";
import { APIError } from "./exceptions";
import { logAPICall, logDebug } from "./utils/logger";

export interface SkillMetadataItem {
  name: string;
  description: string;
}

export interface SkillsMetadataResult {
  skills: SkillMetadataItem[];
  skillsRootPath: string;
}

export interface GetMetadataOptions {
  imageId?: string;
  skillNames?: string[];
}

/**
 * Beta skills service (trial feature).
 */
export class BetaSkillsService {
  private agentBay: AgentBay;

  constructor(agentBay: AgentBay) {
    this.agentBay = agentBay;
  }

  /**
   * Get skills metadata without starting a sandbox.
   *
   * @param options - Optional filtering parameters.
   * @param options.imageId - Image ID to determine the skills root path.
   * @param options.groupIds - Filter by skill group IDs.
   * @returns SkillsMetadataResult with skills list and skillsRootPath.
   */
  async getMetadata(
    options: GetMetadataOptions = {}
  ): Promise<SkillsMetadataResult> {
    const request = new $_client.GetSkillMetaDataRequest({
      authorization: `Bearer ${this.agentBay.getAPIKey()}`,
      imageId: options.imageId,
      skillGroupIds: options.skillNames,
    });

    const maxAttempts = 3;
    let delayMs = 200;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        logAPICall("GetSkillMetaData(beta)");
        const response = await this.agentBay
          .getClient()
          .getSkillMetaData(request);

        if (!response.body || typeof response.body !== "object") {
          throw new APIError(
            "GetSkillMetaData failed: invalid response format"
          );
        }

        if (response.body.success === false) {
          const code = response.body.code ? `[${response.body.code}] ` : "";
          const msg = response.body.message || "Unknown error";
          throw new APIError(`GetSkillMetaData failed: ${code}${msg}`);
        }

        const data = response.body.data;
        if (!data) {
          throw new APIError("GetSkillMetaData failed: missing Data field");
        }

        const skillPath = String(data.skillPath || "");
        const metaDataList = data.metaDataList || [];

        const skills: SkillMetadataItem[] = [];
        for (const raw of metaDataList) {
          const name = String(raw?.name || "").trim();
          if (!name) {
            continue;
          }
          const description = String(raw?.description || "");
          skills.push({ name, description });
        }

        return { skills, skillsRootPath: skillPath };
      } catch (error: unknown) {
        const msg = error instanceof Error ? error.message : `${error}`;
        logDebug(`GetSkillMetaData attempt ${attempt} failed: ${msg}`);
        if (
          attempt < maxAttempts &&
          (msg.includes("ServiceUnavailable") || msg.includes("503"))
        ) {
          await new Promise((resolve) => setTimeout(resolve, delayMs));
          delayMs *= 2;
          continue;
        }
        if (error instanceof APIError) {
          throw error;
        }
        throw new APIError(`GetSkillMetaData failed: ${msg}`);
      }
    }
    throw new APIError("GetSkillMetaData failed: exhausted retries");
  }

  /**
   * @deprecated Use getMetadata() instead.
   */
  async listMetadata(): Promise<SkillMetadataItem[]> {
    const request = new $_client.ListSkillMetaDataRequest({
      authorization: `Bearer ${this.agentBay.getAPIKey()}`,
    });

    const maxAttempts = 3;
    let delayMs = 200;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        logAPICall("ListSkillMetaData(beta)");
        const response = await this.agentBay
          .getClient()
          .listSkillMetaData(request);

        if (!response.body || typeof response.body !== "object") {
          throw new APIError(
            "ListSkillMetaData(beta) failed: invalid response format"
          );
        }

        if (response.body.success === false) {
          const code = response.body.code ? `[${response.body.code}] ` : "";
          const msg = response.body.message || "Unknown error";
          throw new APIError(`ListSkillMetaData(beta) failed: ${code}${msg}`);
        }

        const data = response.body.data;
        if (!Array.isArray(data)) {
          throw new APIError(
            "ListSkillMetaData(beta) failed: invalid Data field"
          );
        }

        const items: SkillMetadataItem[] = [];
        for (const raw of data) {
          const name = String((raw as any)?.name || "").trim();
          if (!name) {
            continue;
          }
          const description = String((raw as any)?.description || "");
          items.push({ name, description });
        }
        return items;
      } catch (error: unknown) {
        const msg = error instanceof Error ? error.message : `${error}`;
        logDebug(`ListSkillMetaData(beta) attempt ${attempt} failed: ${msg}`);
        if (
          attempt < maxAttempts &&
          (msg.includes("ServiceUnavailable") || msg.includes("503"))
        ) {
          await new Promise((resolve) => setTimeout(resolve, delayMs));
          delayMs *= 2;
          continue;
        }
        if (error instanceof APIError) {
          throw error;
        }
        throw new APIError(`ListSkillMetaData(beta) failed: ${msg}`);
      }
    }
    throw new APIError("ListSkillMetaData(beta) failed: exhausted retries");
  }
}
